rlay-backend-neo4j/src/lib.rs

```diff
@@ -144,35 +147,166 @@ impl Neo4jBackend {
+
+        for rel_cid_tuple in &rel_cid_tuples {
+            match empty_entity[rel_cid_tuple.0.as_str()] {
+                Value::Array(_) => {
+                    if !entity[rel_cid_tuple.0.as_str()].is_array() {
+                        entity[rel_cid_tuple.0.as_str()] = Value::Array(Vec::new());
+                    }
+                    entity[rel_cid_tuple.0.as_str()].as_array_mut().unwrap()
+                        .push(rel_cid_tuple.1.as_str().into());
+                }
+                Value::String(_) => {
+                    entity[rel_cid_tuple.0.as_str()] = rel_cid_tuple.1.as_str().into();
+                }
+                Value::Null => {
+                    entity[rel_cid_tuple.0.as_str()] = rel_cid_tuple.1.as_str().into();
+                }
+                _ => unimplemented!(),
+            }
+        }
+
+        let web3_entity: FormatWeb3<Entity> = match serde_json::from_value(entity) {
+            Err(_) => return None,
+            Ok(v) => v
+        };
+        Some(web3_entity.0)
+    }
+
+    /// Convert a "pattern comprehension" object (part of a row - see query for details) to entity
+    fn pattern_object_to_entities(object: &Value) -> Vec<Entity> {
+        let entity = Self::pattern_object_to_entity(object);
+
+        let children: Value = object["children"].clone();
+        let mut children_entities = Self::pattern_objects_to_entities(&children);
+
+        let mut entity_vec: Vec<Entity> = vec![];
+        match entity {
+            None => (),
+            Some(entity) => entity_vec.push(entity)
+        }
+        entity_vec.append(&mut children_entities);
+        entity_vec
+    }
+
+    /// Convert a list of "pattern comprehension" objects to entities
+    /// see usage in pattern_object_to_entities (recursive usage)
+    fn pattern_objects_to_entities(objects: &Value) -> Vec<Entity> {
+        let mut entities: Vec<Entity> = vec![];
+        if objects.is_array() {
+            entities = objects
+                .as_array()
+                .unwrap()
+                .iter()
+                .map(|pattern_object| {
+                    Self::pattern_object_to_entities(pattern_object)
+                })
+            .fold(vec![], |mut all_entities, entities| {
+                all_entities.extend(entities);
+                all_entities
+            });
+        }
+        entities
+    }
+
+    /// Convert "pattern comprehension" rows to resolved entities hash map
+    fn pattern_rows_to_resolve_entities(rows: Rows) -> ResolvedEntities {
+        let mut resolve_entities_map = HashMap::<String, Vec<Entity>>::new();
+        for row in rows {
+            let cid: String = row.get("cid").unwrap();
+            let data: Value = row.get("data").unwrap();
+            resolve_entities_map.insert(
                 cid,
-                retrieved_cid
-            ));
+                Self::pattern_object_to_entities(&data)
+            );
         }
+        resolve_entities_map
+    }
```

Comment:
I think it would make sense to put `pattern_rows` and `pattern_object` logic into a separate module?

What would be a location for that? `rlay-backend-neo4j/src/pattern_comprehension.rs` (or similar file name)?

cc @hobofan

Reply:
=> we leave it in here for now